Use creation date for tagging and pushing Docker images

yvanzo opened the pull request:
# Problem

Breaking changes may occur when updating Python base images.

# Solution

As proposed by @alastair and @mayhem (see https://github.com/metabrainz/docker-python/pull/8#pullrequestreview-526269352 and [example](https://chatlogs.metabrainz.org/brainzbot/metabrainz/msg/4698725/)), a new tagging scheme with creation date would allow to keep using older base images until all dependent repositories have been updated.

New tags with creation date have already been added for current images, see https://hub.docker.com/r/metabrainz/python/tags

This pull request makes `push.sh` to add a tag with creation date. It also handles the case when two images have been pushed on the same day. See commit https://github.com/metabrainz/docker-python/commit/6dd85a8aeb96789790ae3760c8f87ee584c742b1 for details.

alastair commented on push.sh:
```diff
@@ -1,15 +1,59 @@
+		| sed 's/^\(....\)-\(..\)-\(..\)T.*$/\1\2\3/')
+	date_version=${version}-${created}
+	if [[ " ${remote_tags[*]} " == *" ${date_version} "* ]]
+	then
```
it seems like this code exists to do the following:
 - build an image metabrainz/python:3.5
 - find the date that the image was built using docker-inspect
 - retag it to include the date that it was built
 - if this tag already exists, pull that tag, add .0 to the end, and re-push (backup)
 - push the recently built image

I don't think that we're going to have many situations where we build such an image more than once in a day. In fact, we build these images very rarely. 
My plan for this would have been to do either:
 - Automatically get the current date using `date` and add to the `-t` flag of docker build
 - just accept this date as an argument to push.sh, and manage versions manually (as we do in listenbrainz, acousticbrainz)

yvanzo replied:
Already discussed on IRC but here is a summary. This pull request is following Docker Hub’s common practices to update stable tag `x.y` while adding more specific tag  `x.y-date`. Added documentation in commit <https://github.com/metabrainz/docker-python/pull/9/commits/483e4af13dd782ace3d8bbe601c019dcf9de4a8f>.

> I don't think that we're going to have many situations where we build such an image more than once in a day.

We already had the case with `3.6-1`. It is rare but it seems safer to have a plan when it happens.

> Automatically get the current date using `date` and add to the `-t` flag of docker build

Yes, that would work too. But the idea here is to push multiple tags anyway.

> just accept this date as an argument to push.sh

I would avoid that as it makes very easy to enter a wrong date or to override an existing tag using shell history.

alastair reviewed (commented):
My feeling is that this is a bit too complex for what we need. Can you give some more description of why you think we need so much automation related to finding existing versions and performing backups?

If we keep this system, we definitely need some more detailed documentation that explains how it works

yvanzo commented on push.sh:
```diff
@@ -1,15 +1,59 @@
+# shellcheck disable=SC2207
+remote_tags=($(wget -q \
+	https://registry.hub.docker.com/v1/repositories/${image_name}/tags \
+	-O - | jq -r '.[] | .name'))
```
Right, checking for all support programs in commit <https://github.com/metabrainz/docker-python/pull/9/commits/9b813ff0fd8f939b483ae44c19b2dc1f82f2f33f>.

alastair reviewed (approved):
Looks good to me. Should we try and push some images with it before we merge?

yvanzo commented:
I just successfully tested it with `3.6`! Even the additional sequence for the rare case of updating more than once in a day (and fixed that case with commit https://github.com/metabrainz/docker-python/pull/9/commits/255ca6929a1da848e72b840d66553cfad24ab118). I removed the test images after that. I rebased it to apply fIxes. Ready for merge to me.

alastair commented:
thanks!
